src/corehost/cli/fxr/hostfxr.cpp

```diff
@@ -274,6 +276,167 @@ SHARED_API int32_t hostfxr_resolve_sdk(
+//         disallow_prerelease (0x1)
+//           do not allow resolution to return a prerelease SDK version 
+//           unless  prerelease version was specified via global.json.
+//
+//   result
+//      Callback invoked to return the resolved sdk directory and 
+//      global.json path used.
+//      
+//      If resolution succeeds, resolved_sdk_dir will hold the
+//      path to the resolved SDK director, otherwise it will
+//      be null.
+//
+//      If global.json is used, then global_json_path will hold
+//      the path to global.json. If there was global.json found,
+//      or the contents of global.json did not impact resolution
+//      (e.g. no version specified), then global_json_path will
+//      be null.
+//
+// Return value:
+//   0 on success, otherwise failure
+//   0x8000809b - SDK could not be resolved (SdkResolverResolveFailure)
+// 
+// String encoding:
+//   Windows     - UTF-16 (pal::char_t is 2 byte wchar_t)
+//   Unix        - UTF-8  (pal::char_t is 1 byte char)
+//
+SHARED_API int32_t hostfxr_resolve_sdk2(
+    const pal::char_t* exe_dir,
+    const pal::char_t* working_dir,
+    int32_t flags,
+    hostfxr_resolve_sdk2_result_fn result)
+{
+    trace::setup();
+
+    trace::info(_X("--- Invoked hostfxr [commit hash: %s] hostfxr_resolve_sdk2"), _STRINGIFY(REPO_COMMIT_HASH));
+
+    if (exe_dir == nullptr)
+    {
+        exe_dir = _X("");
+    }
+
+    if (working_dir == nullptr)
+    {
+        working_dir = _X("");
+    }
+
+    pal::string_t resolved_sdk_dir;
+    pal::string_t global_json_path;
+
+    bool success = sdk_resolver_t::resolve_sdk_dotnet_path(
+        exe_dir, 
+        working_dir,
+        &resolved_sdk_dir,
+        (flags & hostfxr_resolve_sdk2_flags_t::disallow_prerelease) != 0,
+        &global_json_path);
+
+    assert(success || resolved_sdk_dir.empty());
+
+    result(
+        resolved_sdk_dir.empty() ? nullptr : resolved_sdk_dir.c_str(),
+        global_json_path.empty() ? nullptr : global_json_path.c_str());
+
+    return success
+        ? StatusCode::Success 
+        : StatusCode::SdkResolverResolveFailure;
+}
+
+
+typedef void (*hostfxr_get_available_sdks_result_fn)(
+    int32_t sdk_count,
+    const pal::char_t *sdk_dirs[]);
+
+//
+// Returns the list of all available SDKs ordered by ascending version.
+//
+// Invoked by MSBuild resolver when the latest SDK used without global.json
+// present is incompatible with the current MSBuild version. It will select
+// the compatible SDK that is closest to the end of this list.
+//
+// Parameters:
+//    exe_dir
+//      The path to the dotnet executable.
+//
+//    result
+//      Callback invoke to return the list of SDKs by their directory paths.
+//
+// Return value:
+//   0 on success, otherwise failure
+//
+// String encoding:
+//   Windows     - UTF-16 (pal::char_t is 2 byte wchar_t)
+//   Unix        - UTF-8  (pal::char_t is 1 byte char)
+//
+SHARED_API int32_t hostfxr_get_available_sdks(
```

Comment:
I am intentionally giving Visual Studio the freedom to apply some policy of its own here. I agree with your general principle in the abstract, but there's a trade off at play in this case: which is worse -- teaching tooling  too much about hostfxr internals or teaching hostfxr too much aboout tooling internals? 

The external policy that applies in the MSBuild case that does not apply in the mainline dotnet case is that we have to consider whether MSBuild is too old to use a given SDK:

* When you say `dotnet build`, then we pick a full .NET Core SDK including msbuild itself and the rest of the toolset. There is no incompatibility possible.
* When you say `msbuild`, then we pick only the SDK's msbuild tasks and targets and use the toolset from the calling msbuild. The risk is that the tasks and targets are incompatible with the toolset.

We could technically implement this compat check here, but the layering feels wrong to me. Also, regardless of which layering is more sound in theory, it is cost-effective to keep the things that are specific to the VS case out of this repo and in managed code.

Another issue with moving the code to determine if an SDK is compatible with a given msbuild version here, is that we would we either need to either rediscover the compatibility on each call (parsing the minimumMSBuildVersion file out of each SDK several times over the lifetime of a VS instance), or add some static state to the implementation here.

One thing that has been troubling me is that we don't really need/want to have a formal public surface area for the line between VS and the SDK resolution logic in this repo. The goal is really just to share as much code as possible. In other words, we don't want to change the msbuild resolver and hostfxr every time the policy changes, but I also don't want to be constrained to make further changes if the way we've decided to separate responsibilities ends up being a problem. We don't actually need these API to live in hostfxr.dll. We could consider building them into something else that only ever lives in the msbuild sdk resolver folder and that we do not support from any other caller. @eerhardt and I discussed that and the feeling at the time was that the effort to produce a different library and flow it to the CLI repo wasn't worth it for what started as a single, simple entry point. As the complexity here is increasing, it may be worth reconsidering. In all honesty, I am not all certain that we won't need a hostfxr_resolve_sdk3 at some point.  I think separating these concerns out of hostfxr itself is probably the right direction to go, but there is also still pressure to get this VS functionality in very soon...

Reply:
I see the tradeoffs and I think you hit the nail on the head.  The higher order thing that I think is missing is the set of principles/rules on why we would or would not include a shared_api.  In some sense this API fits really well because I think that having hostfxr be the sole piece of code that understand how to walk our directory structure and return versions numbers feels like.  On the other hand, how it is then used in the CLI feel like we are exposing an internal detail.
You bring up a good point that even though I think of hostfxr as being the 'host' in the VS scenario msbuild/vs are really the host and having them own the policy feels right.
Long way of saying this API seems fine to me.